### PR TITLE
Mobs can climb(and buckle) into seats while laying on the ground

### PR DESCRIPTION
**Pull request by Sirryan2002**
adds all the neccesary code to allow crew members to climb onto seats even if they're incapacitated.

## What Does This PR Do
Crew members who are laying on the floor are now able to climb into any seat that they are under or directly next to. Given that they are not stunned and have at least one hand. They can accomplish this by dragging their sprite onto the seat.

When they attempt to do this, one of three things happen:
- After a delay, they move onto the chair and buckle themselves to it: "
- They fail minorly and their mob still moves to the chair, but they are still on the ground
- They fail majorly and their mob still moves to the chair, they smack their head on the chair, take 15 brute damage to the head,
take 50 stamina damage, take 15 brain damage, and they are knocked out for 2-3 seconds.

If someone has **two hands**, they have a 75% chance of succeeding, a 20% chance of minorly failing, and a 5% chance of majorly failing. It will take them two full seconds in order to complete this action.

if someone has **one hand**, they have 25% chance of succeeding, a 60% chance of minorly failing, and a 15% chance of majorly failing. It will take them five full seconds in order to complete this action.

Messages sent:
"You attempt to climb onto the chair" Attempting
"You fail to climb onto the chair. Come on you got this!" Minor Fail
"SHIT! Your hand slips and you slam your forehead into the chair!" Major Fail
"You pull yourself onto the chair" Success

This has been tested multiple times and it works with all chairs including things like the paramedics trolley. Simple mobs aren't affected.

## Why It's Good For The Game
People who do not have legs(or cannot stand up for whatever reason) and are not stunned should still be able to climb into chairs/wheelchairs. Currently, If you unclip someone who has no legs from a wheelchair/seat they are left completely helpless. This feature would allow those people to have a chance to get back up and in cases where they are injured/in danger access their items (emergency injector, internals, etc). 

This feature, however, does not give a free pass/buff to players who intentionally start off their round as paraplegics. It takes time to climb back into a chair and there's a chance that you can fail or even injure yourself in the process. It makes wheelchairs have a lot more utility while also providing some form of risk.

## Changelog
:cl:
add: Mobs can now climb into chairs if they're incapacitated(but not stunned)
/:cl:

### CODE REVIEW

**Comment by S34NW** [comment]
is your PR description accurate?

`Messages sent:
"You pull yourself onto the chair" Attempting`

**Comment by Sirryan2002** [comment]
> is your PR description accurate?
> 
> `Messages sent: "You pull yourself onto the chair" Attempting`

fixed

**Review by Dumbdumn5 — APPROVED** [review]
Though we don't have full blown crawling yet, I like this as a bit of a flavor change. Especially if it works for rolling beds and shit like the paramed carts. 👍

File: code/game/objects/buckling.dm [hunk 1]
```diff
@@ -117,11 +117,71 @@
 /atom/movable/proc/post_unbuckle_mob(mob/living/M)
 	return
 
+/atom/movable/proc/user_buckle_minorfail(mob/living/M, mob/user)
+	M.forceMove(loc)
+	M.visible_message("<span class='notice'>[M] fails to pull [M.p_them()]self up onto [src].</span>",\
+		"<span class='warning'>You fail to climb onto [src]. Come on you got this!</span>",\
```
Comment (S34NW) [inline code/game/objects/buckling.dm:123]:
```suggestion
		"<span class='warning'>You fail to climb onto [src]. Come on, you've got this!</span>",\
```

File: code/game/objects/buckling.dm [hunk 2]
```diff
@@ -117,11 +117,71 @@
...
+		"<span class='italics'>You hear someone grunting.</span>")
+
+/atom/movable/proc/user_buckle_majorfail(mob/living/M, mob/user) //Knocks user out and does minor brute/brain damage
+	M.StopResting()
+	M.forceMove(loc) //All that effort wasn't for nothing!
+	M.visible_message("<span class='warning'>[M]'s hand slips and they slam their forhead on [src]!</span>",\
```
Comment (S34NW) [inline code/game/objects/buckling.dm:129]:
```suggestion
	M.visible_message("<span class='warning'>[M]'s hand slips and they slam their forehead on [src]!</span>",\
```

File: code/game/objects/buckling.dm [hunk 3]
```diff
@@ -117,11 +117,71 @@
...
+	M.Paralyse(5)
+	M.apply_damage(50, STAMINA)
+	M.apply_damage(15, BRUTE, "head")
+	M.AdjustSleeping(1)
+	playsound(src.loc, 'sound/items/trayhit1.ogg', 50, 0) //bonk
+
+//Allows people to climb onto chairs if they're next to or under them and incapacitated but not stunned
+//especially helpful for people with no legs. However it can fail on random chance based on how many hands they have
+/atom/movable/proc/user_buckle_incapacitated(mob/living/M, mob/user, check_loc = TRUE)
+	M.visible_message("<span class='warning'>[M] looks like theyre attempting to pull [M.p_them()]self up onto [src].</span>",\
```
Comment (S34NW) [inline code/game/objects/buckling.dm:143]:
```suggestion
	M.visible_message("<span class='warning'>[M] looks like they're attempting to pull [M.p_them()]self up onto [src].</span>",\
```

**Review by S34NW — CHANGES_REQUESTED** [review]
Grammar tweaks, everything else is fine grammar wise.

**Comment by MRgaty** [comment]
the chances of failing are pretty high and the punishment for it is way over the top IMO.

File: code/game/objects/buckling.dm [hunk 4]
```diff
@@ -117,11 +117,71 @@
 /atom/movable/proc/post_unbuckle_mob(mob/living/M)
 	return
 
+/atom/movable/proc/user_buckle_minorfail(mob/living/M, mob/user)
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:120]: I can't say I'm a fan of having two separate procs for a minor and major fail. Perhaps you could mimic the severity system EMPs and explosions have, using numbers?

```dm
/atom/movable/proc/user_buckle_fail(mob/living/M, mob/user, severity)
	switch(severity)
		if(1)
			// Minor fail
		if(2)
			// Major fail
		if(3)
			// target.dust()
```

File: code/game/objects/buckling.dm [hunk 5]
```diff
@@ -117,11 +117,71 @@
...
+	M.visible_message("<span class='warning'>[M]'s hand slips and they slam their forehead on [src]!</span>",\
+		"<span class='danger'>SHIT! Your hand slips and you slam your forehead into [src]!</span>",\
+		"<span class='italics'>You hear a thwack and a heavy thump.</span>")
+	M.StartResting() //could if statement these lines down to playsound if you wanted this only to happen with certain seats
+	M.adjustBrainLoss(10)
+	M.Paralyse(5)
+	M.apply_damage(50, STAMINA)
+	M.apply_damage(15, BRUTE, "head")
+	M.AdjustSleeping(1)
+	playsound(src.loc, 'sound/items/trayhit1.ogg', 50, 0) //bonk
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:138]:
```suggestion
	playsound(src, 'sound/items/trayhit1.ogg', 50, FALSE) //bonk
```

In any case, `playsound` plays on the turf of whatever is passed.

File: code/game/objects/buckling.dm [hunk 6]
```diff
@@ -117,11 +117,71 @@
...
+/atom/movable/proc/user_buckle_minorfail(mob/living/M, mob/user)
+	M.forceMove(loc)
+	M.visible_message("<span class='notice'>[M] fails to pull [M.p_them()]self up onto [src].</span>",\
+		"<span class='warning'>You fail to climb onto [src]. Come on, you've got this!</span>",\
+		"<span class='italics'>You hear someone grunting.</span>")
+
+/atom/movable/proc/user_buckle_majorfail(mob/living/M, mob/user) //Knocks user out and does minor brute/brain damage
+	M.StopResting()
+	M.forceMove(loc) //All that effort wasn't for nothing!
+	M.visible_message("<span class='warning'>[M]'s hand slips and they slam their forehead on [src]!</span>",\
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:129]:
```suggestion
	M.visible_message("<span class='warning'>[M]'s hand slips and [M.p_they()] slam [M.p_their()] forehead on [src]!</span>",\
```

File: code/game/objects/buckling.dm [hunk 7]
```diff
@@ -117,11 +117,71 @@
...
 	return
 
+/atom/movable/proc/user_buckle_minorfail(mob/living/M, mob/user)
+	M.forceMove(loc)
+	M.visible_message("<span class='notice'>[M] fails to pull [M.p_them()]self up onto [src].</span>",\
+		"<span class='warning'>You fail to climb onto [src]. Come on, you've got this!</span>",\
+		"<span class='italics'>You hear someone grunting.</span>")
+
+/atom/movable/proc/user_buckle_majorfail(mob/living/M, mob/user) //Knocks user out and does minor brute/brain damage
+	M.StopResting()
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:127]: Curious, why stop resting to immediately rest after? For the animation?

File: code/game/objects/buckling.dm [hunk 8]
```diff
@@ -117,11 +117,71 @@
...
+	M.Paralyse(5)
+	M.apply_damage(50, STAMINA)
+	M.apply_damage(15, BRUTE, "head")
+	M.AdjustSleeping(1)
+	playsound(src.loc, 'sound/items/trayhit1.ogg', 50, 0) //bonk
+
+//Allows people to climb onto chairs if they're next to or under them and incapacitated but not stunned
+//especially helpful for people with no legs. However it can fail on random chance based on how many hands they have
+/atom/movable/proc/user_buckle_incapacitated(mob/living/M, mob/user, check_loc = TRUE)
+	M.visible_message("<span class='warning'>[M] looks like they're attempting to pull [M.p_them()]self up onto [src].</span>",\
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:143]:
```suggestion
	M.visible_message("<span class='warning'>[M] looks like [M.p_they()] [M.p_are()] attempting to pull [M.p_them()]self up onto [src].</span>",\
```

File: code/game/objects/buckling.dm [hunk 9]
```diff
@@ -117,11 +117,71 @@
...
+				return FALSE
+	else if (do_after(user, 5 SECONDS, target = src, use_default_checks=FALSE)) //One hand
+		if(!in_range(user, src) || !isturf(user.loc) || M.anchored)
+			return FALSE
+		if (buckle_chance <= 20)
+			user_buckle_majorfail(M, user)
+			return FALSE
+		if (buckle_chance <= 75)
+			user_buckle_minorfail(M, user)
+			return FALSE
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:166]: Duplicate code!

I would define some variables beforehand, corresponding to the duration and fail chances depending on the user's presence of hands.

```dm
var/buckle_time = 2 SECONDS
var/minor_fail_chance = 5
var/major_fail_chance = 25
if(!user.has_both_hands())
	buckle_time = 5 SECONDS
	minor_fail_chance = 20
	major_fail_chance = 75

if(do_after(user, buckle_time, target = src, use_default_checks = FALSE))
	// ETC...
```

Also, the above numbers should be made into defines at the top of the file, as they are magic numbers.

File: code/game/objects/buckling.dm [hunk 10]
```diff
@@ -117,11 +117,71 @@
...
+	M.forceMove(loc)
+	if (M.get_num_legs()) //people shouldn't go back to resting after unbuckling
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:168]:
```suggestion
	if(M.get_num_legs()) //people shouldn't go back to resting after unbuckling
```

File: code/game/objects/buckling.dm [hunk 11]
```diff
@@ -117,11 +117,71 @@
...
+		M.StopResting()
+	M.visible_message("<span class='notice'>[M] pulls [M.p_them()]self up onto the [src].</span>",\
+		"<span class='notice'>You pull yourself onto the [src].</span>",\
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:171]:
```suggestion
	M.visible_message("<span class='notice'>[M] pulls [M.p_them()]self up onto [src].</span>",\
		"<span class='notice'>You pull yourself onto [src].</span>",\
```

File: code/game/objects/buckling.dm [hunk 12]
```diff
@@ -117,11 +117,71 @@
...
 //Wrapper procs that handle sanity and user feedback
 /atom/movable/proc/user_buckle_mob(mob/living/M, mob/user, check_loc = TRUE)
-	if(!in_range(user, src) || !isturf(user.loc) || user.incapacitated() || M.anchored)
+	if(!in_range(user, src) || !isturf(user.loc) || M.anchored)
 		return FALSE
 
+	if ((user.incapacitated() && !user.IsStunned()) && (user.has_right_hand() || user.has_left_hand()) && M == user)
+		if (!user_buckle_incapacitated(M,user,TRUE))
+			return FALSE
+	else if (user.incapacitated())
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:183]:
```suggestion
	if((user.incapacitated() && !user.IsStunned()) && (user.has_right_hand() || user.has_left_hand()) && M == user)
		if(!user_buckle_incapacitated(M, user, TRUE))
			return FALSE
	else if(user.incapacitated())
```

**Review by dearmochi — CHANGES_REQUESTED** [review]

**Comment by Sirryan2002** [comment]
Alright I've condensed the fails into one proc, moved some magic numbers into #define's and just cleaned up a lot of stuff.

File: code/game/objects/buckling.dm [hunk 13]
```diff
@@ -117,11 +121,63 @@
 /atom/movable/proc/post_unbuckle_mob(mob/living/M)
 	return
 
+/atom/movable/proc/user_buckle_fail(mob/living/M, mob/user, severity)
+	M.forceMove(loc) //All that effort wasn't for nothing!
+	switch(severity)
+		if (1)
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:127]:
```suggestion
		if(1)
```

File: code/game/objects/buckling.dm [hunk 14]
```diff
@@ -117,11 +121,63 @@
...
+			M.visible_message("<span class='notice'>[M] fails to pull [M.p_them()]self up onto [src].</span>",\
+				"<span class='warning'>You fail to climb onto [src]. Come on, you've got this!</span>",\
+				"<span class='italics'>You hear someone grunting.</span>")
+		if (2)
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:131]:
```suggestion
		if(2)
```

File: code/game/objects/buckling.dm [hunk 15]
```diff
@@ -1,3 +1,7 @@
+#define MINOR_FAIL_CHANCE 25
+#define MAJOR_FAIL_CHANCE 5
+#define INCAPACITATED_BUCKLE_TIME 3 SECONDS
+#define ONE_HAND_MULTIPLIER 3
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:4]:
```suggestion
#define ONE_HAND_MULTIPLIER 3

```

File: code/game/objects/buckling.dm [hunk 16]
```diff
@@ -117,11 +121,63 @@
...
+		"<span class='italics'>You hear someone grunting.</span>")
+	var/buckle_chance = rand(1, 100)
+	var/time_multiplier = 1
+	M.StartResting()
+	if(!user.has_both_hands())
+		buckle_chance /= ONE_HAND_MULTIPLIER
+		time_multiplier++
+
+	if(do_after(user, INCAPACITATED_BUCKLE_TIME * time_multiplier, target = src, use_default_checks=FALSE)) //Two hands
+		if (buckle_chance <= MAJOR_FAIL_CHANCE)
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:156]:
```suggestion
		if(buckle_chance <= MAJOR_FAIL_CHANCE)
```

File: code/game/objects/buckling.dm [hunk 17]
```diff
@@ -117,11 +121,63 @@
...
+			user_buckle_fail(M, user, 2)
+			return FALSE
+		if (buckle_chance <= MINOR_FAIL_CHANCE)
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:159]:
```suggestion
		if(buckle_chance <= MINOR_FAIL_CHANCE)
```

File: code/game/objects/buckling.dm [hunk 18]
```diff
@@ -117,11 +121,63 @@
...
+		"<span class='warning'>You attempt to climb onto [src].</span>",\
+		"<span class='italics'>You hear someone grunting.</span>")
+	var/buckle_chance = rand(1, 100)
+	var/time_multiplier = 1
+	M.StartResting()
+	if(!user.has_both_hands())
+		buckle_chance /= ONE_HAND_MULTIPLIER
+		time_multiplier++
+
+	if(do_after(user, INCAPACITATED_BUCKLE_TIME * time_multiplier, target = src, use_default_checks=FALSE)) //Two hands
```
Comment (dearmochi) [inline code/game/objects/buckling.dm:155]:
```suggestion
	if(do_after(user, INCAPACITATED_BUCKLE_TIME * time_multiplier, target = src, use_default_checks = FALSE)) //Two hands
```

**Review by dearmochi — COMMENTED** [review]
A few stylistic nits, mostly about spacing but the code looks fine otherwise

**Review by dearmochi — APPROVED** [review]

**Comment by Sirryan2002** [comment]
> the chances of failing are pretty high and the punishment for it is way over the top IMO.

Sorry I haven't gotten to this, the only time where there's going to be a serious chance of failing is when you have one hand and no legs, otherwise you either have a very low chance of failing or you can just stand up. If my math is right, there's still only a 50% chance that after 4 attempts with the worst conditions you will majorly fail once. The punishment is some light head trauma and a 5-10 second stun. For someone who would otherwise be completely incapacitated without the climbing feature I believe this is a healthy tradeoff. If you really think this is over the top, please drop some values you think would be better.

**Comment by Fox-McCloud** [comment]
This methodology is kinda messy, to be blunt. 

To be perfectly honest, I think I'd rather this be emergently fixed by the upcoming combat rework (which will add crawling and alter buckling, to a degree)---I'd rather not complicate a system that's likely to be ripped out soon, anyway.

**Comment by Sirryan2002** [comment]
One maintainet objection, as much as I like this feature it would probably be more beneficial for the server if I waited for foxes PR

**Pull request closed without merging**